Upgrade Scala 2.13 LTS to 2.13.16

bartoszkosiorek opened the pull request:
Fixes: #1699

### Description
Upgrade Scala 2.13 LTS to 2.13.16

### Motivation
<!-- Mandatory: A summary of why you are making this change. -->

WojciechMazur commented on third_party/repositories/scala_3_6.bzl:
```diff
@@ -205,7 +205,7 @@ artifacts = {
         ],
     },
     "io_bazel_rules_scala_scala_compiler_2": {
-        "artifact": "org.scala-lang:scala-compiler:2.13.15",
+        "artifact": "org.scala-lang:scala-compiler:2.13.16",
         "sha256": "4c200cd193c082bec14a2a2dffe6a1ba5f8130b1b27c79ee54c936dfcafc8ed9",
     },
```
Was it updated using the search and replace? The SHA would probably not match. The files in this directory should be updated using `./scripts/create_repository.py` after updating the Scala 2.13 version specified there.

mbland replied:
Which dependencies weren't updated? If we need to update the script accordingly, we can.

WojciechMazur reviewed (approved):
LGTM

mbland reviewed (commented):
I updated `scripts/create_repository.py` locally to use 2.13.16, and got almost the same result...except that the `io_bazel_rules_scala_scala_compiler{,_2}` entries in this pull request are missing the `deps` field:

```json
    "io_bazel_rules_scala_scala_compiler": {
        "artifact": "org.scala-lang:scala-compiler:2.13.16",
        "sha256": "f59982714591e321ba9c087af2c8666e2f5fb92b11a0cef72c2c5e9b342152d3",
        "deps": [ 
            "@io_bazel_rules_scala_scala_library",
            "@io_bazel_rules_scala_scala_reflect",
            "@io_github_java_diff_utils_java_diff_utils",
            "@org_jline_jline",
        ],
    },
```

The script is set up not to update existing entries, which is why these were never added before (under the "if it ain't broke" principle). But if the scripts adds them on update, and the tests pass, we should be able to trust the new information.

At the same time, by the same principle, the tests _do_ appear to be passing after this update without the new `deps`. So It's not a blocking issue, but I wanted to raise the point that if `scripts/create_repository.py` generates breaking changes, we should update the script; if its changes pass all tests, in general, we should let its output stand.

bartoszkosiorek commented:
> The script is set up not to update existing entries, which is why these were never added before (under the "if it ain't broke" principle). But if the script adds them on update, and the tests pass, we should be able to trust the new information.

Thanks for Review and quality checks.
How did you force script to update existing entries?

mbland commented:
> Thanks for Review and quality checks. How did you force script to update existing entries?

To force an update of an artifact, you could remove its existing entry from the `third_party/repositories/scala_*.bzl` file, and `scripts/create_repository.py` will add it back. (Unless the artifact is marked "testonly"; these artifacts _are_ manually updated, and the script leaves them alone.) You could also artificially set the entry to reference an older artifact version.

---

The script will leave alone any entry that already matches the version resolved by `cs` (executed by `ArtifactResolver._fetch_artifact_data()`), has `"testonly": True` set, or has a version _newer_ than the resolved version. In other words, if the script doesn't see a need to update the artifact version, it won't update it at all.

If it _does_ detect the need to update an artifact's entry, it will also set its `deps` field accordingly, even if it didn't have one before. Hence why `org.scala-lang:scala-compiler:2.13.15` didn't have a `deps` field before (that was the existing version when I update and started running the script), and why `org.scala-lang:scala-compiler:2.13.16` generates one.

---

Now that I think about it, we _could_ add a flag to force an update even for matching versions, to add `deps` information. Or better yet, instead of adding that complexity, we could just delete _all_ the generated entries and recreate them. I haven't felt the need to do this yet, since everything seems to be working as is, but we could.

bartoszkosiorek commented:
> > Thanks for Review and quality checks. How did you force script to update existing entries?
> 
> To force an update of an artifact, you could remove its existing entry from the `third_party/repositories/scala_*.bzl` file, and `scripts/create_repository.py` will add it back. (Unless the artifact is marked "testonly"; these artifacts _are_ manually updated, and the script leaves them alone.) You could also artificially set the entry to reference an older artifact version.

I have updated the dependencies, to be aligned with your results.
Thanks.

In another branch I have also removed all dependencies, and regenerate it from scratch. A lot of dependencies were simply removed. I will create separate Pull Request with such changes.

simuons reviewed (approved):
Thanks, @bartoszkosiorek
